src/services/newMessagesService.ts

```diff
@@ -541,11 +552,80 @@ export default class NewMessagesService {
                   return ResponseErrorStatusNotDefinedInSpec(response);
               }
             })
+          ),
+          TE.chainW((response) =>
+            this.validateThirdPartyMessageResponse(message, response)
           )
         )
       )
     );
 
+  private readonly validateThirdPartyMessageResponse = (
+    message: MessageWithThirdPartyData,
+    response: ThirdPartyMessageDetails
+  ): TE.TaskEither<IResponseErrorBadGateway, ThirdPartyMessage> => {
+    // PN does not need this validation because it is managed completely in a different way with different specs
+    if (message.sender_service_id === PN_SERVICE_ID) {
+      return TE.of(response);
+    }
+    if (
```

Comment:
I'd separate each of these cases and put a comment on top of them to explain the check with common words.